File: tsc/src/add-node.ts
interface Node {
  value: number;
  left: Node | null;
  right: Node | null;
}

export function addNode(root: Node | null, value: number) {
  if (root === null) {
    return { value, left: null, right: null };
  }

  if (value < root.value) {
    if (root.left === null) {
      root.left = { value, left: null, right: null };
    } else {
      addNode(root.left, value);
    }
  } else {
    if (root.right === null) {
      root.right = { value, left: null, right: null };
    } else {
      addNode(root.right, value);
    }
  }

  return root;
}

File: go/add-node.go
package solution

type Node struct {
	value int
	left  *Node
	right *Node
}

func addNode(root *Node, value int) *Node {
	if root == nil {
		return &Node{value, nil, nil}
	}

	if value < root.value {
		if root.left == nil {
			root.left = &Node{value, nil, nil}
		} else {
			addNode(root.left, value)
		}
	} else {
		if root.right == nil {
			root.right = &Node{value, nil, nil}
		} else {
			addNode(root.right, value)
		}
	}

	return root
}
